common/scripts/babel-conditional-preprocess.js

```diff
@@ -63,62 +65,68 @@ exports.default = babelHelper.declare((_api, opts) => {
-        Handle(path, annotations);
+        Handle(path, match, featurePrefixes);
       },
 
       ClassMethod(path) {
-        Handle(path, annotations);
+        Handle(path, match, featurePrefixes);
       },
 
       ClassProperty(path) {
-        Handle(path, annotations);
+        Handle(path, match, featurePrefixes);
       },
     }
   };
 });
 
-function Handle(path, annotations) {
+function Handle(path, match, featurePrefixes) {
   let { node } = path;
-  let removed = false;
-
-  for (const annotation of annotations) {
-    const {
-      match
-    } = annotation;
-
-    if (node.leadingComments && node.leadingComments.length > 0) {
-      for (const comment of node.leadingComments) {
-        if (comment.value.includes(match)) {
-          removed = true;
-        }
-        if (removed) {
-          comment.ignore = true;
-          // Comment will be inherit to next line even it is set to 'ignore'
-          // this will cause some unexpected removals of code
-          // clear the comment to ensure nothing gets wrong
-          comment.value = '';
-        }
-      }
-    }
 
-    if (removed) {
-      // We cannot remove Expression in JSXExpressionContainer cause it is not correct for AST
-      // Replacing it with jSXEmptyExpression will get us the same result
-      // There will always be only one expression under JSXExpressionContainer
-      if (path?.container?.type === 'JSXExpressionContainer') {
-        path.replaceWith(t.jSXEmptyExpression());
-      } else {
-        path.remove();
-      }
+  const shouldRemove = node.leadingComments && node.leadingComments.some((comment) => containsDirective(comment, match, featurePrefixes));
+  if (!shouldRemove) {
+    return;
+  }
+
+  node.leadingComments && node.leadingComments.forEach((comment) => {
+    if (!containsDirective(comment, match, featurePrefixes)) {
+      return;
     }
+    comment.ignore = true;
+    // Comment is inherited by next line even it is set to 'ignore'.
+    // Clear the conditional compilation directive to avoid removing the
+    // next line.
+    comment.value = '';
+  })
+
+  // We cannot remove Expression in JSXExpressionContainer cause it is not correct for AST
+  // Replacing it with jSXEmptyExpression will get us the same result
+  // There will always be only one expression under JSXExpressionContainer
+  if (path?.container?.type === 'JSXExpressionContainer') {
+    path.replaceWith(t.jSXEmptyExpression());
+  } else {
+    path.remove();
   }
 }
 
+function containsDirective(comment, match, featurePrefixes) {
+  if (comment.value.includes(match)) {
+    // legacy annotation
+    return true;
+  }
+  // Check for partial prefix first to avoid checking each feature in most cases.
+  if (!comment.value.includes(FEATURE_PREFIX)) {
+    return false;
+  }
+  if(featurePrefixes.some((match) => comment.value.includes(match))) {
+    return true;
+  }
+  throw new Error(`Unknown conditional compilation feature in: ${comment.value}`);
```

Comment:
Devs would benefit from actual filename and line number.
Possible to add that to the error message?

Reply:
I tried... but babel doesn't seem to be populating `node.loc.filename`

@PorterNan  do you know why?

In my latest commit, the filename is always `undefined`.